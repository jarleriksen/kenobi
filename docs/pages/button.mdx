---
name: Button
route: /button
---

import { Playground } from 'docz';
import styled from 'styled-components';
import Button from '../../src/button/src/Button.tsx';
import PurpleButton from '../components/PurpleButton';
import Row from '../components/Row';
import Component from '@reach/component-component';

# Button

### `Button` is a base component of Kenobi. It can contain a label and an icon before or after the label.

## Usage

<Playground>
  <Button>Base button</Button>
</Playground>

## Variations

**Component and Row are not a part of Kenobi**

<Playground>
  <Component initialState={{ isLoading: false }}>
    {({ state, setState }) => (
      <React.Fragment>
        <div style={{ marginBottom: '12px' }}>
          <Button
            onClick={() => setState({ isLoading: !state.isLoading })}
            marginBottom={4}>
            {state.isLoading ? 'Hide' : 'Show'} loading state
          </Button>
        </div>
        <Row>
          <Button marginRight={8} isLoading={state.isLoading}>
            Default
          </Button>
          <Button intent="primary" marginRight={8} isLoading={state.isLoading}>
            Primary
          </Button>
          <Button intent="success" marginRight={8} isLoading={state.isLoading}>
            Success
          </Button>
          <Button intent="warning" marginRight={8} isLoading={state.isLoading}>
            Warning
          </Button>
          <Button intent="danger" marginRight={8} isLoading={state.isLoading}>
            Danger
          </Button>
          <Button marginRight={8} iconBefore="edit" isLoading={state.isLoading}>
            Icon before
          </Button>
          <Button marginRight={8} iconAfter="edit" isLoading={state.isLoading}>
            Icon after
          </Button>
        </Row>
      </React.Fragment>
    )}
  </Component>
</Playground>

[More information on usage of Icons](/icon)

## Sizing

Default size for buttons are 32px.
`Button` will handle its own scaling based on the height.

<Playground>
  <Row>
    <Button height={24} marginRight={8}>
      Height 24
    </Button>
    <Button height={32} marginRight={8}>
      Height 32
    </Button>
    <Button height={40} marginRight={8}>
      Height 40
    </Button>
    <Button height={48} marginRight={8}>
      Height 48
    </Button>
    <Button height={56} marginRight={8}>
      Height 56
    </Button>
  </Row>
</Playground>

## Usage as a link

Due to using Styled-components as our CSS-in-JS library it is really simple to switch out the underlying
HTML element. We just have to pass the prop **isLink** to the Button component.
The button will pass the **as** prop to the Styled element.

This will render the button as an **anchor** element.

<Playground>
  <Button isLink={true} href="/button#button-as-a-link">
    I am a link!
  </Button>
</Playground>

## Extending styles

It's very easy to extend Button styles with styled-components. (Haven't tested any other solutions).

<PurpleButton>I am now Purple!</PurpleButton>

```jsx
import styled from 'styled-components';
import { Button } from 'kenobi-ui';

const PurpleButton = styled(Button)`
  background-color: #551a8b;
`;

<PurpleButton>I am now Purple!</PurpleButton>;
```
