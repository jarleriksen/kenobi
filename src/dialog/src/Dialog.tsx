import React, { useContext } from 'react';
import { Overlay } from '../../overlay';
import styled from 'styled-components';
import { ThemeContext } from '../../theme/ThemeConsumer';
import { Heading, Paragraph } from '../../typography';
import { Button } from '../../button';
import { Icon } from '../../icon';
import { VisuallyHidden } from '../../visually-hidden';
import DialogWrapper from './DialogWrapper';

const ANIMATION_DURATION = 200;

interface DefaultProps {
  topOffset?: any;
  sideOffset?: any;
  minHeightContent?: number;
}

export interface DialogProps extends DefaultProps {
  /**
   * Controls if the Drawer is open or closed.
   */
  isOpen: boolean;

  /**
   * Dialog content
   */
  children: React.ReactNode;

  /**
   * If the dialog should show a header
   */
  showHeader?: boolean;

  /**
   * If showHeader is true, show this title.
   * Will be used as aria-label if no custom label is given.
   */
  dialogTitle?: string;

  /**
   * If set the aria-label will be set to this.
   * Otherwise use dialogTitle or simply nothing.
   */
  ariaLabel?: string;

  /**
   * Whether or not to show close icon
   */
  hasClose?: boolean;

  /**
   * If clicking the overlay should close the Dialog.
   */
  shouldCloseOnOverlayClick?: boolean;

  /**
   * If the Dialog should close on Escape press.
   */
  shouldCloseOnEsc?: boolean;

  /**
   * Function to control the value of isOpen.
   *
   * This is a user provided function.
   */
  onRequestClose: () => void;

  /**
   * Width of the Dialog
   */
  width?: number;
}

const DialogContentWrapper = styled.div<{ minHeight?: number }>`
  display: flex;
  overflow: auto;
  flex-direction: column;
  min-height: ${({ minHeight }) => minHeight}px;

  box-shadow: ${({ theme }) =>
    `0 0 1px ${theme.colors.palette.neutral.base}, 0 16px 24px -8px ${theme.colors.palette.neutral.dark}`};
`;

const DialogHeader = styled.div`
  display: flex;
  align-items: center;
  border-bottom: 1px solid ${({ theme }) => theme.colors.palette.neutral.light};
  padding: 16px;
  flex-shrink: 0;
`;

const DialogCloseIconWrapper = styled.div`
  border-radius: 100%;
  width: 32px;

  transition: background-color 200ms ease 0s;
`;

const DialogContent = styled.div`
  padding: 16px;
  overflow: auto;
`;

const Dialog = ({
  isOpen = false,
  showHeader = true,
  hasClose = true,
  shouldCloseOnOverlayClick = true,
  shouldCloseOnEsc = true,
  width = 560,
  topOffset = '12vmin',
  sideOffset = '16px',
  minHeightContent = 80,
  ...props
}: DialogProps) => {
  const themeContext = useContext(ThemeContext);

  const { children, dialogTitle, ariaLabel, onRequestClose } = props;

  const sideOffsetWithUnit = Number.isInteger(sideOffset)
    ? `${sideOffset}px`
    : sideOffset;
  const maxWidth = `calc(100% - ${sideOffsetWithUnit} * 2)`;

  const topOffsetWithUnit = Number.isInteger(topOffset)
    ? `${topOffset}px`
    : topOffset;
  const maxHeight = `calc(100% - ${topOffsetWithUnit} * 2)`;

  const dialogContent =
    typeof children === 'string' ? (
      <Paragraph size={400}>{children}</Paragraph>
    ) : (
      children
    );

  return (
    <Overlay
      isOpen={isOpen}
      onRequestClose={onRequestClose}
      shouldCloseOnClick={shouldCloseOnOverlayClick}
      shouldCloseOnEsc={shouldCloseOnEsc}
      childrenWrapperStyles={{
        display: 'flex',
        alignItems: 'flex-start',
        justifyContent: 'center',
      }}>
      {({ state }) => (
        <DialogWrapper
          role="dialog"
          aria-modal="true"
          aria-labelledby={ariaLabel || dialogTitle || ''}
          theme={themeContext}
          width={width}
          maxWidth={maxWidth}
          maxHeight={maxHeight}
          marginX={sideOffsetWithUnit}
          marginY={topOffsetWithUnit}
          state={state}
          animationDuration={ANIMATION_DURATION}
          data-testid="kenobi-dialog">
          <DialogContentWrapper
            theme={themeContext}
            minHeight={minHeightContent}>
            {showHeader && (
              <DialogHeader theme={themeContext}>
                <Heading as="h4" size={600}>
                  {dialogTitle}
                </Heading>
                {hasClose && (
                  <DialogCloseIconWrapper>
                    <Button
                      tabIndex={0}
                      intent="none"
                      onClick={onRequestClose}
                      data-testid="kenobi-dialog-close-button">
                      <VisuallyHidden>Close</VisuallyHidden>
                      <Icon aria-hidden icon="cross" />
                    </Button>
                  </DialogCloseIconWrapper>
                )}
              </DialogHeader>
            )}

            <DialogContent>{dialogContent}</DialogContent>
          </DialogContentWrapper>
        </DialogWrapper>
      )}
    </Overlay>
  );
};

export default Dialog;
