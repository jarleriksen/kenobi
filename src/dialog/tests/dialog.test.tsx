import React from 'react';
import { render } from '@testing-library/react';
import Dialog from '../src/Dialog';
import { Button } from '../../button';
import '@testing-library/jest-dom/extend-expect';

describe('Dialog', () => {
  test('should not be in portal when closed', () => {
    const { getByTestId, container, asFragment } = render(
      <Dialog
        isOpen={false}
        dialogTitle="Dialog title"
        onRequestClose={() => {}}>
        <div title="insideDialog" />
      </Dialog>,
    );

    const portal = getByTestId('kenobi-portal');
    const dialog: HTMLElement | null = document.querySelector(
      '[data-testid="kenobi-dialog"]',
    );

    expect(portal).not.toContainElement(dialog);
    expect(asFragment()).toMatchSnapshot();
  });

  test('should be in portal when open', () => {
    const { getByTestId, container, asFragment } = render(
      <Dialog
        isOpen={true}
        dialogTitle="Dialog title"
        onRequestClose={() => {}}>
        <div title="insideDialog" />
      </Dialog>,
    );

    const portal = getByTestId('kenobi-portal');
    const dialog: HTMLElement | null = document.querySelector(
      '[data-testid="kenobi-dialog"]',
    );

    expect(portal).toContainElement(dialog);
    expect(asFragment()).toMatchSnapshot();
  });

  test('close button has focus when first opened', () => {
    const { getByTestId, asFragment } = render(
      <Dialog
        isOpen={true}
        dialogTitle="Dialog title"
        onRequestClose={() => {}}>
        <div title="insideDialog" />
      </Dialog>,
    );

    expect(getByTestId('kenobi-dialog-close-button')).toHaveFocus();
    expect(asFragment()).toMatchSnapshot();
  });

  test('can focus another element', () => {
    const { getByTestId, asFragment } = render(
      <Dialog
        isOpen={true}
        dialogTitle="Dialog title"
        onRequestClose={() => {}}>
        <Button data-autofocus>Hello</Button>
      </Dialog>,
    );

    expect(getByTestId('kenobi-button')).toHaveFocus();
    expect(asFragment()).toMatchSnapshot();
  });
});
