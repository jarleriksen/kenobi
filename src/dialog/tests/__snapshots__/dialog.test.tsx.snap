// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialog can focus another element 1`] = `<DocumentFragment />`;

exports[`Dialog close button has focus when first opened 1`] = `<DocumentFragment />`;

exports[`Dialog should be in portal when open 1`] = `<DocumentFragment />`;

exports[`Dialog should not be in portal when closed 1`] = `<DocumentFragment />`;
